agent/fw_cakephp.c

```diff
@@ -324,17 +307,22 @@ void nr_cakephp_enable_1(TSRMLS_D) {
 }
 
 /*
- * Enable CakePHP 2.0+
+ * Enable CakePHP 4.0+
  */
-void nr_cakephp_enable_2(TSRMLS_D) {
-  nr_php_wrap_user_function(NR_PSTR("Controller::invokeAction"),
-                            nr_cakephp_name_the_wt_2 TSRMLS_CC);
+void nr_cakephp_enable(TSRMLS_D) {
+  nr_php_wrap_user_function(
+      NR_PSTR("Cake\\Controller\\Controller::invokeAction"),
+      nr_cakephp_name_the_wt_4 TSRMLS_CC);
+  nr_txn_suggest_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                               PHP_PACKAGE_VERSION_UNKNOWN);
 #if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
     && !defined OVERWRITE_ZEND_EXECUTE_DATA
   nr_php_wrap_user_function_before_after_clean(
-      NR_PSTR("CakeException::__construct"), nr_cakephp_problem_2, NULL, NULL);
+      NR_PSTR("Cake\\Core\\Exception\\CakeException::__construct"),
+      nr_cakephp_problem_4, NULL, NULL);
```

Comment:
I'm not sure this is needed. When Observer API is used to hook into Zend Engine, the agent should do the right thing and record the error on the transaction even though framework installs its own exception handler. See Yii's exception instrumentation [here](https://github.com/newrelic/newrelic-php-agent/blob/R11.3/agent/fw_yii.c#L197-L224) for reference.

Reply:
I have experimented with the above and noticed that the error is not recorded as a traced error. Instead, I have added code that retrieves the exception caught by CakePHP's error handler and reports it as a traced error: [86108b5](https://github.com/newrelic/newrelic-php-agent/pull/983/commits/86108b52af6eaade06b1bbd578f703edf78c56e4)

Reply:
Leaving information for posterity:
- [CakePHP 4.x Error & Exception Handling](https://book.cakephp.org/4/en/development/errors.html)
- [CakePHP 5.x Error & Exception Handling](https://book.cakephp.org/5/en/development/errors.html)

The agent correctly detects that CakePHP registers its default exception handler `Cake\Error\ExceptionTrap::handleException` ([CakePHP 4.x](https://github.com/cakephp/cakephp/blob/4.x/src/Error/ExceptionTrap.php#L186), [CakePHP 5.x](https://github.com/cakephp/cakephp/blob/5.x/src/Error/ExceptionTrap.php#L173)). However, that exception handler is used only for uncaught exceptions. In a CakePHP application, exceptions thrown in controller actions are caught in `Cake\Error\Middleware\ErrorHandlerMiddleware::process` method ([CakePHP 4.x](https://github.com/cakephp/cakephp/blob/5.x/src/Error/Middleware/ErrorHandlerMiddleware.php#L105-L121), [CakePHP 5.x](https://github.com/cakephp/cakephp/blob/5.x/src/Error/Middleware/ErrorHandlerMiddleware.php#L105-L121)). `Cake\Error\Middleware\ErrorHandlerMiddleware` is part of CakePHP application's default middleware stack ([CakePHP 4.x](https://github.com/cakephp/app/blob/4.x/src/Application.php#L80), [CakePHP 5.x](https://github.com/cakephp/app/blob/5.x/src/Application.php#L70)) so wrapping `Cake\Error\Middleware\ErrorHandlerMiddleware::handleException` is the right approach for the agent to record error events for exceptions thrown by controller actions in standard CakePHP applications.